datastore/gcloud/rest/datastore/query.py

```diff
@@ -0,0 +1,220 @@
+from typing import Any  # pylint: disable=unused-import
+from typing import Dict  # pylint: disable=unused-import
+from typing import List  # pylint: disable=unused-import
+
+from gcloud.rest.datastore.constants import MoreResultsType
+from gcloud.rest.datastore.constants import ResultType
+from gcloud.rest.datastore.entity import EntityResult
+from gcloud.rest.datastore.filter import Filter
+from gcloud.rest.datastore.property_order import PropertyOrder
+from gcloud.rest.datastore.value import Value
+
+
+class BaseQuery(object):
+    json_key = ''
+    value_kind = Value
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self.to_repr())
+
+    @classmethod
+    def from_repr(cls, data):
+        # type: (Dict[str, Any]) -> BaseQuery
+        raise NotImplementedError
+
+    def to_repr(self):
+        # type: () -> Dict[str, Any]
+        raise NotImplementedError
+
+
+# https://cloud.google.com/datastore/docs/reference/data/rest/v1/projects/runQuery#Query
+class Query(BaseQuery):
+    json_key = 'query'
+
+    # TODO: support `projection` and `distinctOn`
+    def __init__(self,
+                 kind='',            # type: str
+                 query_filter=None,  # type: Filter
```

Comment:
This should be `Optional[Filter]` -- could you go through the type hints and make sure you've marked down which are nullable?

Reply:
I'll admit I mindlessly converted the Python 3 annotations - gcloud-aio is wrong as well. Will fix